File: src/main/java/com/matag/cards/ability/AbilityService.java
package com.matag.cards.ability;

import com.matag.cards.ability.type.AbilityType;
import com.matag.cards.properties.PowerToughness;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static com.matag.language.StringUtils.replaceLast;
import static java.util.Collections.singletonList;

@Component
public class AbilityService {
  // permanent
  private static final String DAMAGE = "DAMAGE:";
  private static final String CONTROLLER_DAMAGE = "CONTROLLER_DAMAGE:";
  private static final String TAPPED = ":TAPPED";
  private static final String UNTAPPED = ":UNTAPPED";
  private static final String DOES_NOT_UNTAP_NEXT_TURN = ":DOES_NOT_UNTAP_NEXT_TURN";
  private static final String DESTROYED = ":DESTROYED";
  private static final String RETURN_TO_OWNER_HAND = ":RETURN_TO_OWNER_HAND";
  private static final String CONTROLLED = ":CONTROLLED";
  private static final String PLUS_1_COUNTERS = "PLUS_1_COUNTERS:";
  private static final String MINUS_1_COUNTERS = "MINUS_1_COUNTERS:";
  private static final String KEYWORD_COUNTER = "KEYWORD_COUNTER:";
  // player
  private static final String DRAW = "DRAW:";
  private static final String LIFE = "LIFE:";

  public PowerToughness powerToughnessFromParameters(List<String> parameters) {
    return parameters.stream()
      .filter(parameter -> parameter.contains("/"))
      .map(PowerToughness::powerToughness)
      .findFirst()
      .orElse(PowerToughness.powerToughness("0/0"));
  }

  public PowerToughness powerToughnessFromParameter(String parameter) {
    return powerToughnessFromParameters(singletonList(parameter));
  }

  public int damageFromParameter(String parameter) {
    return getParameterIntValue(parameter, DAMAGE);
  }

  public int controllerDamageFromParameter(String parameter) {
    return getParameterIntValue(parameter, CONTROLLER_DAMAGE);
  }

  public boolean tappedFromParameter(String parameter) {
    return parameter.equals(TAPPED);
  }

  public boolean untappedFromParameter(String parameter) {
    return parameter.equals(UNTAPPED);
  }

  public boolean doesNotUntapNextTurnFromParameter(String parameter) {
    return parameter.equals(DOES_NOT_UNTAP_NEXT_TURN);
  }

  public boolean destroyedFromParameter(String parameter) {
    return parameter.equals(DESTROYED);
  }

  public boolean returnToOwnerHandFromParameter(String parameter) {
    return parameter.equals(RETURN_TO_OWNER_HAND);
  }

  public boolean controlledFromParameter(String parameter) {
    return parameter.equals(CONTROLLED);
  }

  public int plus1CountersFromParameter(String parameter) {
    return getParameterIntValue(parameter, PLUS_1_COUNTERS);
  }

  public int minus1CountersFromParameter(String parameter) {
    return getParameterIntValue(parameter, MINUS_1_COUNTERS);
  }

  public AbilityType keywordCounterFromParameter(String parameter) {
    return getParameterValue(parameter, KEYWORD_COUNTER)
        .map(AbilityType::valueOf)
        .orElse(null);
  }

  public int drawFromParameter(String parameter) {
    return getParameterIntValue(parameter, DRAW);
  }

  public int lifeFromParameter(String parameter) {
    return getParameterIntValue(parameter, LIFE);
  }

  public String parametersAsString(List<String> parameters) {
    var text = parameters.stream().map(this::safeParameterAsString).collect(Collectors.joining(", "));
    return replaceLast(text, ",", " and");
  }

  public String safeParameterAsString(String parameter) {
    try {
      return parameterAsString(parameter);
    } catch (Exception e) {
      return parameter.toLowerCase();
    }
  }

  public String parameterAsString(String parameter) throws RuntimeException {
    if (parameter.contains("/")) {
      return parameter;

    } else if (parameter.startsWith("DAMAGE:")) {
      return damageFromParameter(parameter) + " damage";

    } else if (parameter.startsWith("CONTROLLER_DAMAGE:")) {
      return "to its controller " + controllerDamageFromParameter(parameter) + " damage";

    } else if (tappedFromParameter(parameter)) {
      return "tapped";

    } else if (untappedFromParameter(parameter)) {
      return "untapped";

    } else if (doesNotUntapNextTurnFromParameter(parameter)) {
      return "doesn't untap next turn";

    } else if (destroyedFromParameter(parameter)) {
      return "destroyed";

    } else if (returnToOwnerHandFromParameter(parameter)) {
      return "returned to its owner's hand";

    } else if (controlledFromParameter(parameter)) {
      return "controlled";

    } else if (parameter.startsWith("PLUS_1_COUNTERS:")) {
      return plus1CountersFromParameter(parameter) + " +1/+1 counters";

    } else if (parameter.startsWith("MINUS_1_COUNTERS:")) {
      return minus1CountersFromParameter(parameter) + " -1/-1 counters";

    } else if (parameter.startsWith("KEYWORD_COUNTER:")) {
      return "a " + abilityParameterAsString(keywordCounterFromParameter(parameter)) + " counter";

    } else if (parameter.startsWith("LIFE:")) {
      var life = lifeFromParameter(parameter);
      return (life > 0 ? "gain " + life : "lose " + (-life)) + " life";

    } else if (parameter.startsWith("DRAW:")) {
      var draw = drawFromParameter(parameter);
      return "draw " + draw + " card" + (draw > 1 ? "s" : "");

    } else {
      return abilityParameterAsString(AbilityType.valueOf(parameter));
    }
  }

  private String abilityParameterAsString(AbilityType abilityType) {
    return abilityType.getText().substring(0, abilityType.getText().length() - 1).toLowerCase();
  }

  private Optional<String> getParameterValue(String parameter, String parameterType) {
    if (parameter.startsWith(parameterType)) {
      return Optional.of(parameter.replace(parameterType, ""));
    }
    return Optional.empty();
  }

  private int getParameterIntValue(String parameter, String parameterType) {
    return getParameterValue(parameter, parameterType).map(Integer::parseInt).orElse(0);
  }
}

File: src/main/kotlin/com/matag/cards/ability/AbilityService.kt
package com.matag.cards.ability

import com.matag.cards.ability.type.AbilityType
import com.matag.cards.properties.PowerToughness
import com.matag.language.replaceLast
import org.springframework.stereotype.Component
import java.util.*
import kotlin.math.abs

@Component
class AbilityService {

    fun powerToughnessFromParameters(parameters: List<String>) =
        parameters
            .filter { it.contains("/") }
            .map { it.split("/") }
            .map { PowerToughness(it[0].toInt(), it[1].toInt()) }
            .firstOrNull() ?: PowerToughness(0, 0)

    fun powerToughnessFromParameter(parameter: String) =
        powerToughnessFromParameters(listOf(parameter))

    fun damageFromParameter(parameter: String) =
        getParameterIntValue(parameter, DAMAGE)

    fun controllerDamageFromParameter(parameter: String) =
        getParameterIntValue(parameter, CONTROLLER_DAMAGE)

    fun tappedFromParameter(parameter: String) =
        parameter == TAPPED

    fun untappedFromParameter(parameter: String) =
        parameter == UNTAPPED

    fun doesNotUntapNextTurnFromParameter(parameter: String) =
        parameter == DOES_NOT_UNTAP_NEXT_TURN

    fun destroyedFromParameter(parameter: String) =
        parameter == DESTROYED

    fun returnToOwnerHandFromParameter(parameter: String) =
        parameter == RETURN_TO_OWNER_HAND

    fun controlledFromParameter(parameter: String) =
        parameter == CONTROLLED

    fun cancelledFromParameter(parameter: String) =
        parameter == CANCELLED

    fun plus1CountersFromParameter(parameter: String) =
        getParameterIntValue(parameter, PLUS_1_COUNTERS)

    fun minus1CountersFromParameter(parameter: String) =
        getParameterIntValue(parameter, MINUS_1_COUNTERS)

    fun keywordCounterFromParameter(parameter: String) =
        getParameterValue(parameter, KEYWORD_COUNTER)?.let { AbilityType.valueOf(it) }

    fun drawFromParameter(parameter: String) =
        getParameterIntValue(parameter, DRAW)

    fun lifeFromParameter(parameter: String) =
        getParameterIntValue(parameter, LIFE)

    fun scryFromParameter(parameter: String) =
        getParameterIntValue(parameter, SCRY)

    fun parametersAsString(parameters: List<String>) =
        parameters.joinToString(", ") { parameterAsString(it) }
            .replaceLast(",", " and")

    private fun parameterAsString(parameter: String): String {
        if (parameter.contains("/")) {
            return parameter
        } else if (parameter.startsWith("DAMAGE:")) {
            return damageFromParameter(parameter).toString() + " damage"
        } else if (parameter.startsWith("CONTROLLER_DAMAGE:")) {
            return "to its controller " + controllerDamageFromParameter(parameter) + " damage"
        } else if (tappedFromParameter(parameter)) {
            return "tapped"
        } else if (untappedFromParameter(parameter)) {
            return "untapped"
        } else if (doesNotUntapNextTurnFromParameter(parameter)) {
            return "doesn't untap next turn"
        } else if (destroyedFromParameter(parameter)) {
            return "destroyed"
        } else if (returnToOwnerHandFromParameter(parameter)) {
            return "returned to its owner's hand"
        } else if (controlledFromParameter(parameter)) {
            return "controlled"
        } else if (cancelledFromParameter(parameter)) {
            return "cancelled"
        } else if (parameter.startsWith("PLUS_1_COUNTERS:")) {
            return plus1CountersFromParameter(parameter).toString() + " +1/+1 counters"
        } else if (parameter.startsWith("MINUS_1_COUNTERS:")) {
            return minus1CountersFromParameter(parameter).toString() + " -1/-1 counters"
        } else if (parameter.startsWith("KEYWORD_COUNTER:")) {
            return "a " + abilityParameterAsString(keywordCounterFromParameter(parameter)!!) + " counter"
        } else if (parameter.startsWith("LIFE:")) {
            val life = lifeFromParameter(parameter)
            return (if (life > 0) "gain" else "lose") + " ${abs(life)} life"
        } else if (parameter.startsWith("DRAW:")) {
            val draw = drawFromParameter(parameter)
            return "draw " + draw + " card" + (if (draw > 1) "s" else "")
        } else if (parameter.startsWith("SCRY:")) {
            return "scry ${scryFromParameter(parameter)}"
        } else {
            return abilityParameterAsString(AbilityType.valueOf(parameter))
        }
    }

    private fun abilityParameterAsString(abilityType: AbilityType) =
        abilityType.text.dropLast(1).lowercase(Locale.getDefault())

    private fun getParameterValue(parameter: String, parameterType: String) =
        if (parameter.startsWith(parameterType)) {
            parameter.replace(parameterType, "")
        } else {
            null
        }

    private fun getParameterIntValue(parameter: String, parameterType: String) =
        getParameterValue(parameter, parameterType)?.toInt() ?: 0

    companion object {
        private const val DAMAGE = "DAMAGE:"
        private const val CONTROLLER_DAMAGE = "CONTROLLER_DAMAGE:"
        private const val TAPPED = ":TAPPED"
        private const val UNTAPPED = ":UNTAPPED"
        private const val DOES_NOT_UNTAP_NEXT_TURN = ":DOES_NOT_UNTAP_NEXT_TURN"
        private const val DESTROYED = ":DESTROYED"
        private const val RETURN_TO_OWNER_HAND = ":RETURN_TO_OWNER_HAND"
        private const val CONTROLLED = ":CONTROLLED"
        private const val CANCELLED = ":CANCELLED"
        private const val PLUS_1_COUNTERS = "PLUS_1_COUNTERS:"
        private const val MINUS_1_COUNTERS = "MINUS_1_COUNTERS:"
        private const val KEYWORD_COUNTER = "KEYWORD_COUNTER:"
        private const val DRAW = "DRAW:"
        private const val LIFE = "LIFE:"
        private const val SCRY = "SCRY:"
    }
}
